core/job/service/job_service.go

```diff
@@ -1999,3 +1993,143 @@ func (j *JobService) getDownstreamJobs(ctx context.Context, projectName tenant.P
 	}
 	return downstreamJobs, me.ToErr()
 }
+
+func (j *JobService) BulkDeleteJobs(ctx context.Context, projectName tenant.ProjectName, jobsToDelete []*dto.JobToDeleteRequest) (map[string]dto.BulkDeleteTracker, error) {
+	// initiate deletion status per jobs to be deleted
+	deletionTrackerMap := map[string]dto.BulkDeleteTracker{}
+
+	toDeleteJobs := job.Jobs{}
+	for _, jobToDelete := range jobsToDelete {
+		existingJob, err := j.jobRepo.GetByJobName(ctx, projectName, jobToDelete.JobName)
+		if err != nil && !errors.IsErrorType(err, errors.ErrNotFound) {
+			j.logger.Error("error getting existing job %s: %s", jobToDelete.JobName.String(), err)
+			return deletionTrackerMap, err
+		}
+
+		toDeleteJobs = append(toDeleteJobs, existingJob)
+	}
+
+	toDeleteMap := toDeleteJobs.GetNameMap()
+	toDeleteSpecMap := toDeleteJobs.GetFullNameToSpecMap()
+
+	alreadyDeleted := map[job.FullName]bool{}
+	deletedJobs := job.Jobs{}
+	for _, toDeleteJob := range toDeleteJobs {
```

Comment:
can we make the job deletion check in upfront before actually delete it? if we have set of job deletion `S` and it's not safe to delete, and we have subset of `S`, eg. `S1` and it's safe to delete, then job deletion can be partially done (`S1`), but not whole(`S`). as a consequences, I believe we can't retry the pipeline without manual intervention

Reply:
After separate discussion, we've decided to separate the steps to:
1. validation, and
2. the actual deletion.
This require code refactor, will timebox this until 2PM. Else, will go forward with the current approach